pilot/pkg/config/kube/file/controller.go

```diff
@@ -0,0 +1,163 @@
+// Copyright Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package file
+
+import (
+	"fmt"
+
+	"istio.io/istio/pilot/pkg/config/kube/crd"
+	"istio.io/istio/pilot/pkg/model"
+	"istio.io/istio/pkg/config"
+	"istio.io/istio/pkg/config/schema/collection"
+	"istio.io/istio/pkg/config/schema/collections"
+	"istio.io/istio/pkg/kube/krt"
+	krtfiles "istio.io/istio/pkg/kube/krt/files"
+)
+
+var errUnsupportedOp = fmt.Errorf("unsupported operation: the file config store is a read-only view")
+
+type Controller struct {
+	collections map[config.GroupVersionKind]krt.Collection[config.Config]
+	schemas     collection.Schemas
+	handlers    map[config.GroupVersionKind][]krt.HandlerRegistration
+	stop        chan struct{}
+}
+
+func NewController(fileDir string, domainSuffix string, schemas collection.Schemas) (*Controller, error) {
+	stop := make(chan struct{})
+	// TODO: options.KrtDebugger
+	opts := krt.NewOptionsBuilder(stop, "file-monitor", krt.GlobalDebugHandler)
+	watch, err := krtfiles.NewFolderWatch(fileDir, func(b []byte) ([]*config.Config, error) {
+		return parseInputs(b, domainSuffix)
+	}, stop)
+	if err != nil {
+		return nil, err
+	}
+
+	enabledSchemas := make(map[config.GroupVersionKind]bool)
+	for _, s := range schemas.All() {
+		if _, ok := collections.Pilot.FindByGroupVersionKind(s.GroupVersionKind()); ok {
+			enabledSchemas[s.GroupVersionKind()] = true
+		}
+	}
+
+	collections := make(map[config.GroupVersionKind]krt.Collection[config.Config])
+	for s := range enabledSchemas {
+		collections[s] = krtfiles.NewFileCollection(watch, func(c *config.Config) *config.Config {
+			if c.GroupVersionKind == s {
+				return c
+			}
+
+			return nil
+		}, opts.WithName("FileMonitor")...)
+	}
+
+	return &Controller{
+		collections: collections,
+		schemas:     schemas,
+		stop:        stop,
+		handlers:    make(map[config.GroupVersionKind][]krt.HandlerRegistration),
+	}, nil
+}
+
+func (c *Controller) Schemas() collection.Schemas {
+	return c.schemas
+}
+
+func (c *Controller) Get(typ config.GroupVersionKind, name, namespace string) *config.Config {
+	return nil
```

Comment:
Hmm, this seems like we might need to support it.. TBH might be a bug that we don't in the Ingress/Gateway/etc one as well, not sure everywhere Get is used

Reply:
Will check and fix